{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "91eca9cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "#genic regions specific allele read depth\n",
    "infile = '../../../bsa_by_rnaseq/05_mapping/allele_readdepth_genic.txt'\n",
    "winsize = 18000000\n",
    "outfile = '../../../bsa_by_rnaseq/05_mapping/allele_freq_ratio_winsize_' + str(winsize) + '.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "a2029222",
   "metadata": {},
   "outputs": [],
   "source": [
    "windict = {}\n",
    "chromdict = {}\n",
    "idict = {}\n",
    "\n",
    "fi = open(infile)\n",
    "\n",
    "cchrom = ''\n",
    "i = 0\n",
    "\n",
    "for line in fi:\n",
    "    if line[0] == '#':\n",
    "        continue\n",
    "    \n",
    "    counts = line.rstrip().split('\\t')\n",
    "    \n",
    "    chrom = counts.pop(0)\n",
    "    if chrom != cchrom:\n",
    "        cchrom = chrom\n",
    "        start = 0\n",
    "        end = start + winsize -1\n",
    "        i = i + 1\n",
    "        idict[i] = [start, end]\n",
    "        windict[i] = [0,0,0,0,0,0,0,0]\n",
    "        chromdict[i] = chrom\n",
    "    \n",
    "    pos = int(counts.pop(0))\n",
    "    ref = counts.pop(0)\n",
    "    alt = counts.pop(0)\n",
    "    #print(chrom, counts)\n",
    "    \n",
    "    if pos >= start and pos < end:\n",
    "        for x in range(8):\n",
    "            windict[i][x] = windict[i][x] + int(counts[x])\n",
    "            \n",
    "        #print(chrom, pos, counts)\n",
    "        #print(i, chrom, pos, windict[i])\n",
    "\n",
    "    if pos >= end:\n",
    "        start = end + 1\n",
    "        end = start + winsize -1\n",
    "        i = i + 1\n",
    "        idict[i] = [start, end]\n",
    "        windict[i] = [0,0,0,0,0,0,0,0]\n",
    "        chromdict[i] = chrom\n",
    "        \n",
    "        for x in range(8):\n",
    "            windict[i][x] = windict[i][x] + int(counts[x])\n",
    "\n",
    "        #print(chrom, pos, counts)\n",
    "        #print(i, chrom, pos, windict[i])\n",
    "    \n",
    "fi.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "ed431c3e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "313 Chr29 0 17999999 1.0266592159378423 1.1306318297194975 1.0247130340632675 1.154484445030948\n",
      "314 Chr29 18000000 35999999 1.067513785788043 1.11625600810291 1.0736598070597647 1.1450071498078598\n",
      "315 Chr29 36000000 53999999 1.2174965242869955 1.294739881130691 1.2654534967042754 1.3779581258951101\n",
      "316 Chr29 54000000 71999999 1.4911380177913558 1.50631901343704 1.5822813465430914 1.6522181816167787\n",
      "317 Chr29 72000000 89999999 2.6766867849743265 3.0708770694053995 2.758678902395878 3.2222896291485883\n",
      "318 Chr29 90000000 107999999 4.400976782717007 7.223130841121496 7.1234234170398505 7.513362445414848\n",
      "319 Chr29 108000000 125999999 4.159253254522829 6.7854253017817925 5.866270417213813 4.777255476970847\n",
      "320 Chr29 126000000 143999999 2.675933415585187 3.153165735173319 3.097864939149279 3.0207948960911564\n",
      "321 Chr29 144000000 161999999 2.1830109949266943 1.8098869167700022 2.125670519849524 1.7055627217026765\n"
     ]
    }
   ],
   "source": [
    "fo = open(outfile, 'w')\n",
    "fo.write('chromosome\\tstart\\tend\\twindow')\n",
    "fo.write('\\tH14_ref_depth\\tH14_alt_depth\\tM14_ref_depth\\tM14_alt_depth')\n",
    "fo.write('\\tH19_ref_depth\\tH19_alt_depth\\tM19_ref_depth\\tM19_alt_depth')\n",
    "\n",
    "fo.write('\\tH14_ref_freq\\tH14_alt_freq\\tM14_ref_freq\\tM14_alt_freq')\n",
    "fo.write('\\ther14_ref_ratio\\ther14_alt_ratio')\n",
    "fo.write('\\tH19_ref_freq\\tH19_alt_freq\\tM19_ref_freq\\tM19_alt_freq')\n",
    "fo.write('\\ther19_ref_ratio\\ther19_alt_ratio\\n')\n",
    "\n",
    "for i in windict:\n",
    "    #print(i, chromdict[i], windict[i])\n",
    "    fo.write(chromdict[i] + '\\t' + str(idict[i][0]) + '\\t' + str(idict[i][1]) + '\\t' + str(i))\n",
    "    \n",
    "    pseudocounts = [x+1 for x in windict[i]]\n",
    "    [her14He_ref_count, her14He_alt_count, her14Ma_ref_count, her14Ma_alt_count, her19He_ref_count, her19He_alt_count, her19Ma_ref_count, her19Ma_alt_count] = pseudocounts\n",
    "    \n",
    "    #if i == 1:\n",
    "    #    print(pseudocounts)\n",
    "    \n",
    "    her14He_alt_frq = her14He_alt_count / (her14He_alt_count + her14He_ref_count)\n",
    "    her14He_ref_frq = her14He_ref_count / (her14He_alt_count + her14He_ref_count)\n",
    "    her14Ma_alt_frq = her14Ma_alt_count / (her14Ma_alt_count + her14Ma_ref_count)\n",
    "    her14Ma_ref_frq = her14Ma_ref_count / (her14Ma_alt_count + her14Ma_ref_count)\n",
    "    her14_freqs = [her14He_ref_frq, her14He_alt_frq, her14Ma_ref_frq, her14Ma_alt_frq]\n",
    "    her14_ref_ratio = her14He_ref_frq / her14Ma_ref_frq\n",
    "    her14_alt_ratio = her14Ma_alt_frq / her14He_alt_frq\n",
    "    \n",
    "    her19He_alt_frq = her19He_alt_count / (her19He_alt_count + her19He_ref_count)\n",
    "    her19He_ref_frq = her19He_ref_count / (her19He_alt_count + her19He_ref_count)\n",
    "    her19Ma_alt_frq = her19Ma_alt_count / (her19Ma_alt_count + her19Ma_ref_count)\n",
    "    her19Ma_ref_frq = her19Ma_ref_count / (her19Ma_alt_count + her19Ma_ref_count)\n",
    "    her19_freqs = [her19He_ref_frq, her19He_alt_frq, her19Ma_ref_frq, her19Ma_alt_frq]\n",
    "    her19_ref_ratio = her19He_ref_frq / her19Ma_ref_frq\n",
    "    her19_alt_ratio = her19Ma_alt_frq / her19He_alt_frq\n",
    "\n",
    "    #if i == 1:\n",
    "    #    print(her14_freqs, her14_ref_ratio, her14_alt_ratio)\n",
    "    \n",
    "    if chromdict[i] == 'Chr29':\n",
    "        print(i, chromdict[i], idict[i][0], idict[i][1], her14_ref_ratio, her19_ref_ratio, her14_alt_ratio, her19_alt_ratio)\n",
    "    \n",
    "    for x in windict[i]:\n",
    "        fo.write('\\t' + str(x))\n",
    "        \n",
    "    for x in her14_freqs:\n",
    "        fo.write('\\t' + str(x))\n",
    "\n",
    "    fo.write('\\t' + str(her14_ref_ratio) + '\\t' + str(her14_alt_ratio))\n",
    "    \n",
    "    for x in her19_freqs:\n",
    "        fo.write('\\t' + str(x))\n",
    "    \n",
    "    fo.write('\\t' + str(her19_ref_ratio) + '\\t' + str(her19_alt_ratio))\n",
    "\n",
    "    fo.write('\\n')    \n",
    "fo.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a866dda3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d781b10",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (My env.genomics Kernel)",
   "language": "python",
   "name": "env.genomics"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
