{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4b9b5b3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.stats import hypergeom\n",
    "import statsmodels.stats.multitest as multi\n",
    "import pandas as pd\n",
    "import os\n",
    "import glob"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4fcbc009",
   "metadata": {},
   "outputs": [],
   "source": [
    "degfiles = ['../../../her7/10_gene_exp/gene_expression_jen_local/10_gene_exp/burow_her19_v_herm_de_results.txt',\n",
    "            '../../../her7/10_gene_exp/gene_expression_jen_local/10_gene_exp/burow_male_v_her19_de_results.txt',\n",
    "            '../../../her7/10_gene_exp/gene_expression_jen_local/10_gene_exp/burow_male_v_herm_de_results.txt']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8f7f4808",
   "metadata": {},
   "outputs": [],
   "source": [
    "annfile = '../../../her7/02_functional_annotations/Ceratopteris_KEGG_annotations_by_phylogeny.txt'\n",
    "outfile = '../../../her7/10_gene_exp/functional_enrichment_results_KEGG.txt'\n",
    "keggfile = '../../../../dbs/kegg/KEGG_KO_ALL_PARENTS_Jun-30-2024.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "bfe0fd0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# min/max thresholds for tests\n",
    "highCount = 10000000\n",
    "lowCount = 0\n",
    "\n",
    "de_lfc_thres = 0.58\n",
    "de_padj_thres = 0.1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "bc130c87",
   "metadata": {},
   "outputs": [],
   "source": [
    "###################################\n",
    "### Initialize set dictionaries ###\n",
    "###################################\n",
    "degsets = {}\n",
    "\n",
    "for degfile in degfiles:\n",
    "    #print(degfile)\n",
    "    comparison = degfile.split('burow_')[1].split('_de_results')[0]\n",
    "    #print(comparison)\n",
    "    degsets[comparison + '_up'] = set()\n",
    "    degsets[comparison + '_down'] = set()\n",
    "    \n",
    "    df = pd.read_csv(degfile, sep='\\t', index_col=0)\n",
    "    df.index.name = 'gene_id'\n",
    "    df\n",
    "    \n",
    "    # Filter rows based on conditions\n",
    "    de_up_df = df[(df['log2FoldChange'] > de_lfc_thres) & (df['padj'] < de_padj_thres)]\n",
    "    de_down_df = df[(df['log2FoldChange'] < de_lfc_thres * -1) & (df['padj'] < de_padj_thres)]\n",
    "\n",
    "    # Extract gene_id column as Python list\n",
    "    for gene in de_up_df.index.tolist():\n",
    "        #print(gene)\n",
    "        degsets[comparison + '_up'].add(gene)\n",
    "        \n",
    "    for gene in de_down_df.index.tolist():\n",
    "        #print(gene)\n",
    "        degsets[comparison + '_down'].add(gene)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "813dc389",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of genes in each DEG set:\n",
      "her19_v_herm_up 209\n",
      "her19_v_herm_down 1982\n",
      "male_v_her19_up 6222\n",
      "male_v_her19_down 5568\n",
      "male_v_herm_up 4868\n",
      "male_v_herm_down 4885\n"
     ]
    }
   ],
   "source": [
    "# initialize remaining data structures\n",
    "isaDict = {}\n",
    "annotate_line = []\n",
    "keggDict = {}\n",
    "codeDict = {}\n",
    "#codeDict['set'] = {}\n",
    "codeDict['total'] = {}\n",
    "lociDict = {}\n",
    "#lociDict['set'] = set()\n",
    "lociDict['total'] = set()\n",
    "\n",
    "print('Number of genes in each DEG set:')\n",
    "for degset in degsets:\n",
    "    print(degset, len(degsets[degset]))\n",
    "    codeDict[degset] = {}\n",
    "    lociDict[degset] = set()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "3cff47eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "#############################################\n",
    "### Load KEGG relationship hash from file ###\n",
    "#############################################\n",
    "\n",
    "fi = open(keggfile)\n",
    "\n",
    "for line in fi:\n",
    "    currentParents = set()\n",
    "    line = line.rstrip().split('\\t')\n",
    "    currentID = line.pop(0).split(':')[0]\n",
    "    \n",
    "    for parent in line:\n",
    "        currentParents.add(parent)\n",
    "        \n",
    "    #print(currentID, currentParents)\n",
    "    isaDict[currentID] = currentParents\n",
    "\n",
    "fi.close()\n",
    "\n",
    "descfiles = os.path.realpath(keggfile)\n",
    "descfiles = os.path.split(descfiles)[0] + '/*DESC_Jun-30-2024.txt'\n",
    "\n",
    "for infile in glob.glob(descfiles):\n",
    "    #print(infile)\n",
    "    fi = open(infile)\n",
    "    \n",
    "    for line in fi:\n",
    "        term, desc = line.rstrip().split('\\t')\n",
    "        #print(term, desc)\n",
    "        keggDict[term] = desc\n",
    "    \n",
    "    fi.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ec048e5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "######################################\n",
    "### Parse KEGGs in Annotation File ###\n",
    "######################################\n",
    "\n",
    "fi = open(annfile)\n",
    "\n",
    "for line in fi:\n",
    "    locus, kegg = line.rstrip().split('\\t')\n",
    "    keggSet = set()\n",
    "        \n",
    "    keggSet.add(kegg)\n",
    "    if kegg in isaDict:\n",
    "        returnedparents = isaDict[kegg]\n",
    "\n",
    "        for parent in returnedparents:\n",
    "            #skip if parent is a Module and not a full pathway\n",
    "            #regexp = re.compile(r'^M\\d+$')\n",
    "            #if regexp.search(parent):\n",
    "            #    continue\n",
    "            keggSet.add(parent)\n",
    "\n",
    "        #if locus == 'Ceric.1Z203700':\n",
    "            #print(locus,keggSet)\n",
    "    \n",
    "    for kegg in keggSet:\n",
    "        if kegg not in codeDict['total']:\n",
    "            codeDict['total'][kegg] = set()\n",
    "\n",
    "        codeDict['total'][kegg].add(locus)\n",
    "        lociDict['total'].add(locus)    \n",
    "\n",
    "        for degset in degsets:\n",
    "            #print(degset)\n",
    "            if locus in degsets[degset]:\n",
    "                #if locus == 'Ceric.1Z203700':\n",
    "                    #print(degset,locus,keggSet)\n",
    "    \n",
    "                if kegg not in codeDict[degset]:\n",
    "                    codeDict[degset][kegg] = set()\n",
    "\n",
    "                codeDict[degset][kegg].add(locus)\n",
    "                lociDict[degset].add(locus)\n",
    "\n",
    "fi.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "889dd0c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "####################################\n",
    "### Perform hypergeometric tests ###\n",
    "####################################\n",
    "\n",
    "df = pd.DataFrame(columns=['set','kegg','desc','x','N','n','M','pval','genelist'])\n",
    "\n",
    "for degset in degsets:\n",
    "    for kegg in codeDict[degset]:\n",
    "\n",
    "        count = len(codeDict[degset][kegg])\n",
    "        if count > highCount or count < lowCount:\n",
    "            continue\n",
    "\n",
    "        desc = ''\n",
    "        if kegg in keggDict:\n",
    "            desc = keggDict[kegg]\n",
    "\n",
    "        # x is still the number of drawn \"successes\" (ie no. genes in set and in go category)\n",
    "        x = len(codeDict[degset][kegg])\n",
    "        genelist = ', '.join(codeDict[degset][kegg])\n",
    "\n",
    "        # N is the sample size (ie no. genes in set)\n",
    "        N = len(lociDict[degset])\n",
    "\n",
    "        # n is the number of successes in the population (ie no. genes in go category)\n",
    "        n = len(codeDict['total'][kegg])\n",
    "\n",
    "        # M is the population size (ie no. genes total)\n",
    "        M = len(lociDict['total'])\n",
    "\n",
    "        # https://alexlenail.medium.com/understanding-and-implementing-the-hypergeometric-test-in-python-a7db688a7458\n",
    "        # https://github.com/jdrudolph/goenrich\n",
    "        pval = hypergeom.sf(x-1, M, n, N)\n",
    "        df.loc[len(df.index)] = [degset,kegg,desc,x,N,n,M,pval,genelist]  \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "a878c5ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "#########################################\n",
    "### Adjust pvalues for multiple tests ###\n",
    "#########################################\n",
    "if len(df['pval'].tolist()) > 0:\n",
    "    adjpval = multi.multipletests(df['pval'].tolist(), alpha=0.05, method='fdr_bh', is_sorted=False, returnsorted=False)[1]\n",
    "    df['adjpval'] = adjpval\n",
    "    df['cat'] = 'kegg'\n",
    "    df['seqfreq'] = df['x'] / df['N']\n",
    "    df['totalfreq'] = df['n'] / df['M']\n",
    "\n",
    "    df = df[['set','kegg','cat','desc','x','N','seqfreq','n','M','totalfreq','pval','adjpval','genelist']]\n",
    "\n",
    "    df.to_csv(outfile, sep='\\t', index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71910302",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (My env.genomics Kernel)",
   "language": "python",
   "name": "env.genomics"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
