{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d1653a54",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "35f7bf6f",
   "metadata": {},
   "outputs": [],
   "source": [
    "vepfile = '../../../bsa_by_rnaseq/06_vep/vep_output/HER19_variant_effects'\n",
    "samplesfile = '../../../bsa_by_rnaseq/06_vep/samples.txt'\n",
    "outfile = '../../../bsa_by_rnaseq/06_vep/filtered_vep_output.txt'\n",
    "\n",
    "target_chrom = 'Chr29'\n",
    "target_start = 96000000\n",
    "target_end = 112000000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "4c2dca79",
   "metadata": {},
   "outputs": [],
   "source": [
    "snpDict = {}\n",
    "\n",
    "fi = open(vepfile)\n",
    "\n",
    "for line in fi:\n",
    "    if line[0] == '#':\n",
    "        continue\n",
    "        \n",
    "    chrom = line.split('_')[0]\n",
    "    loc = int(line.split('_')[1])\n",
    "    ref = line.split('_')[2].split('/')[0]\n",
    "    alt = line.split('/')[1].split('\\t')[0]\n",
    "    gene = line.split('\\t')[3]\n",
    "    effect  = line.split('\\t')[6]\n",
    "    aa = line.split('\\t')[10]\n",
    "    codon = line.split('\\t')[11]\n",
    "    impact = line.rstrip().split('IMPACT=')[1].split(';')[0]\n",
    "\n",
    "    if chrom == target_chrom and loc > target_start and loc < target_end:\n",
    "        #print(chrom,loc,ref,alt,gene,effect,aa,codon,impact)\n",
    "        snpDict[(chrom, loc)] = (chrom,loc,ref,alt,gene,effect,aa,codon,impact)\n",
    "    \n",
    "fi.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "b89538c7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "46"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(snpDict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "41a9202b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "her19 Chr29 100538618 T C 1 260 0.9961685823754789\n",
      "her19 Chr29 100538619 A C 1 235 0.9873949579831933\n",
      "her19 Chr29 100538627 T G 1 260 0.9961685823754789\n",
      "her19 Chr29 100538629 C G 1 260 0.9961685823754789\n",
      "her19 Chr29 100538630 A C 1 260 0.9961685823754789\n",
      "her19 Chr29 100538632 T C 1 260 0.9961685823754789\n",
      "her19 Chr29 100538634 A G 1 260 0.9961685823754789\n",
      "her19 Chr29 100538638 A T 1 260 0.9961685823754789\n",
      "her19 Chr29 100538640 A G 1 260 0.9961685823754789\n",
      "her19 Chr29 100538642 A T 1 260 0.9961685823754789\n",
      "her19 Chr29 100538659 G A 1 224 0.9955555555555555\n",
      "her19 Chr29 100538675 A C 1 197 0.985\n",
      "her19 Chr29 100538679 C A 1 196 0.9849246231155779\n",
      "her19 Chr29 100538688 C T 1 165 0.9821428571428571\n",
      "her19 Chr29 100538693 T G 1 61 0.9384615384615385\n",
      "her19 Chr29 100538702 C G 1 32 0.9411764705882353\n",
      "her19 Chr29 100538703 C A 1 26 0.9285714285714286\n",
      "her19 Chr29 100538705 T A 1 24 0.9230769230769231\n",
      "her19 Chr29 100538709 C A 1 18 0.9\n",
      "her19 Chr29 101567637 G T 1 61 0.8243243243243243\n",
      "her19 Chr29 103038769 T A 1 253 0.933579335793358\n",
      "her19 Chr29 103038773 T A 1 253 0.9301470588235294\n",
      "her19 Chr29 103038784 A T 1 252 0.9298892988929889\n",
      "her19 Chr29 103038795 C G 1 251 0.9194139194139194\n",
      "her19 Chr29 103038796 G A 1 251 0.916058394160584\n",
      "her19 Chr29 103224204 A G 1 47 1.0\n",
      "her19 Chr29 106588383 G A 1 2 1.0\n",
      "her19 Chr29 106588384 C A 1 2 1.0\n",
      "her19 Chr29 109605065 G C 1 8 1.0\n",
      "her19 Chr29 109605067 A T 1 8 1.0\n",
      "her19 Chr29 109605103 A T 1 8 1.0\n",
      "her19 Chr29 109606046 A C 1 2 1.0\n",
      "her19 Chr29 109622486 G T 1 2 1.0\n",
      "her19 Chr29 109622492 G C 1 2 1.0\n",
      "her19 Chr29 109622774 A C 1 2 1.0\n",
      "her19 Chr29 109622777 T C 1 2 1.0\n",
      "her19 Chr29 109635438 G A 1 2 1.0\n",
      "her19 Chr29 109635455 C T 1 2 1.0\n",
      "her19 Chr29 109635456 A G 1 2 1.0\n",
      "her19 Chr29 109635627 C A 1 2 1.0\n",
      "her19 Chr29 109635628 C A 1 2 1.0\n",
      "her19 Chr29 109635650 T C 1 2 1.0\n",
      "her19 Chr29 109635667 C A 1 2 1.0\n",
      "her19 Chr29 109658631 G A 1 2 1.0\n",
      "her19 Chr29 109658636 G T 1 1 1.0\n",
      "her19 Chr29 109898885 C A 1 2 1.0\n",
      "N8 Chr29 109605065 G C 1/1 1 1.0\n",
      "Hnn Chr29 106588383 G A 1/1 4 1.0\n",
      "Hnn Chr29 106588384 C A 1/1 4 1.0\n",
      "her19xN8ma Chr29 100538618 T *,C 1/2 4 0.26666666666666666\n",
      "her19xN8ma Chr29 100538630 A *,G 1/2 4 0.26666666666666666\n",
      "her19xN8ma Chr29 100538638 A *,ACTCTGATGTC 1/2 4 0.26666666666666666\n",
      "her19xN8ma Chr29 100538642 A *,G 1/2 4 0.3076923076923077\n",
      "her19xN8ma Chr29 100538659 G *,GACAGGAAAGGTA 1/2 4 0.36363636363636365\n"
     ]
    }
   ],
   "source": [
    "sampleDict = {}\n",
    "\n",
    "fi = open(samplesfile)\n",
    "\n",
    "for line in fi:\n",
    "    sample, vcfile = line.rstrip().split('\\t')\n",
    "    sampleDict[sample] = {}\n",
    "    \n",
    "    fi = gzip.open(vcfile, 'rt')\n",
    "\n",
    "    for line in fi:\n",
    "        #print(line)\n",
    "\n",
    "        if line[0] == '#': \n",
    "            continue\n",
    "\n",
    "        col = line.rstrip().split('\\t')\n",
    "        chrom = col[0]\n",
    "        pos = int(col[1])\n",
    "        ref = col[3]\n",
    "        alt = col[4]\n",
    "\n",
    "        #if chrom == target_chrom and pos > target_start and pos < target_end:\n",
    "            #print(chrom,pos,ref,alt)\n",
    "\n",
    "        genotype = col[9].split(':')[0]\n",
    "        allele_depth = col[9].split(':')[1]\n",
    "\n",
    "        if (chrom, pos) in snpDict:\n",
    "            #print(chrom,pos,ref,alt,genotype,allele_depth)\n",
    "            ref_depth = 0\n",
    "            alt_dept = 0\n",
    "\n",
    "            if genotype == '0/0' or genotype == '0':\n",
    "                if ref == snpDict[(chrom, pos)][0] and alt == '.':\n",
    "                    #print(chrom, pos, ref, alt, genotype, allele_depth)\n",
    "                    ref_depth = int(allele_depth)\n",
    "                else:\n",
    "                    #print(chrom, pos, ref, alt, genotype, allele_depth)\n",
    "                    continue\n",
    "\n",
    "            elif genotype == '0/1' or genotype == '1/1' or genotype == '1':\n",
    "                if ref != snpDict[(chrom, pos)][2]:\n",
    "                    continue\n",
    "                if alt == snpDict[(chrom, pos)][3]:\n",
    "                    #print(chrom, pos, ref, alt, genotype, allele_depth)\n",
    "                    ref_depth = int(allele_depth.split(',')[0])\n",
    "                    alt_depth = int(allele_depth.split(',')[1])\n",
    "                else:\n",
    "                    ref_depth = int(allele_depth.split(',')[0]) + int(allele_depth.split(',')[1])\n",
    "                    #print('Warning check ref and alt alleles: ', sample, chrom,pos,ref,alt,genotype,allele_depth, snpDict[(chrom, pos)][2], snpDict[(chrom, pos)][3])\n",
    "                    continue\n",
    "\n",
    "            elif genotype == '1/2':\n",
    "                for depth in allele_depth.split(','):\n",
    "                    ref_depth = ref_depth + int(depth)\n",
    "                \n",
    "                if sample == 'HER19xN8ma' and pos == 100538618:\n",
    "                    ref_depth = 3\n",
    "                    alt_depth = 8\n",
    "                \n",
    "                #print(sample, chrom, pos, ref, alt, genotype, allele_depth, snpDict[(chrom, pos)][2], snpDict[(chrom, pos)][3], ref_depth)\n",
    "\n",
    "            else:\n",
    "                #print('Warning check genotype format:', sample, chrom, pos, ref, alt, genotype, allele_depth, snpDict[(chrom, pos)][2], snpDict[(chrom, pos)][3])\n",
    "                continue\n",
    "\n",
    "            total_depth = ref_depth + alt_depth\n",
    "            alt_prop = 'na'\n",
    "\n",
    "            if total_depth > 0:\n",
    "                alt_prop = alt_depth / total_depth            \n",
    "\n",
    "            sampleDict[sample][(chrom, pos)] = [alt_depth, alt_prop]\n",
    "\n",
    "            print(sample, chrom, pos, ref, alt, genotype, alt_depth, alt_prop)\n",
    "\n",
    "    fi.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "606cecde",
   "metadata": {},
   "outputs": [],
   "source": [
    "fo = open(outfile, 'w')\n",
    "\n",
    "fo.write('chrom\\tlocation\\tref\\talt\\tgene\\tconsequence\\tamino_acids\\tcodons\\timpact')\n",
    "for sample in sampleDict:\n",
    "    fo.write('\\t' + sample)\n",
    "fo.write('\\n')\n",
    "\n",
    "for snp in snpDict:\n",
    "    chrom,loc,ref,alt,gene,effect,aa,codon,impact = snpDict[snp]\n",
    "    fo.write(chrom + '\\t' + str(loc) + '\\t' + ref + '\\t' + alt + '\\t' + gene + '\\t' + effect + '\\t' + aa + '\\t' + codon + '\\t' + impact)\n",
    "    \n",
    "    for sample in sampleDict:\n",
    "        alt_prop = 0\n",
    "        if snp in sampleDict[sample]:\n",
    "            alt_prop = sampleDict[sample][snp][1]\n",
    "\n",
    "        \n",
    "        fo.write('\\t' + str(alt_prop))\n",
    "    fo.write('\\n')\n",
    "            \n",
    "fo.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb118cd0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5cd9dfa7",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (My env.genomics Kernel)",
   "language": "python",
   "name": "env.genomics"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
