{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3dbde609",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "os.environ['QT_QPA_PLATFORM']='offscreen'\n",
    "\n",
    "import Bio\n",
    "from Bio import Seq, SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c9764475",
   "metadata": {},
   "outputs": [],
   "source": [
    "import ete3\n",
    "from ete3 import Tree, faces, TreeStyle, PhyloTree, NodeStyle, TextFace, AttrFace, SeqMotifFace"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "b36a52ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "treefile = '/depot/jwisecav/data/ceratopteris/her7/08_phylogeny/HER7_simple.tre'\n",
    "alnfile = '/depot/jwisecav/data/ceratopteris/her7/08_phylogeny/HER7.aln'\n",
    "speciesfile = '/depot/jwisecav/data/ceratopteris/her7/08_phylogeny/scripts/species_ids.txt'\n",
    "colorfile = '/depot/jwisecav/data/ceratopteris/her7/08_phylogeny/scripts/lineage_colors.dmp'\n",
    "branch_min = 95"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "606e7fa2",
   "metadata": {},
   "outputs": [],
   "source": [
    "seqDict = {}\n",
    "alnLen = 0\n",
    "for record in SeqIO.parse(alnfile, \"fasta\"):\n",
    "    name = record.id\n",
    "    sequence = str(record.seq)\n",
    "    alnLen = len(sequence)\n",
    "    seqDict[name] = sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "50325261",
   "metadata": {},
   "outputs": [],
   "source": [
    "code2name = {}\n",
    "code2lineage = {}\n",
    "fi = open(speciesfile)\n",
    "\n",
    "for line in fi:\n",
    "    if line[0] == '#':\n",
    "        continue\n",
    "    name, spid, lineage = line.rstrip().split('\\t')\n",
    "    #print(name, spid, lineage)\n",
    "    code2name[spid] = name\n",
    "    \n",
    "    code2lineage[spid] = lineage\n",
    "\n",
    "fi.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "1f785c68",
   "metadata": {},
   "outputs": [],
   "source": [
    "lin2color = {}\n",
    "linorder = {}\n",
    "fi = open(colorfile)\n",
    "\n",
    "for line in fi:\n",
    "    if line[0] == '#':\n",
    "        continue\n",
    "\n",
    "    order, color_desc, color, cat, lineage, taxids = line.rstrip().split('\\t')  \n",
    "    order = int(order)\n",
    "    lin2color[lineage] = color\n",
    "    linorder[order] = lineage\n",
    "\n",
    "fi.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "8eac69c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add species and lineage information to all leaves\n",
    "# color target gene red \n",
    "# color all ME034 gene green\n",
    "\n",
    "#t = PhyloTree(treefile, alignment=alnfile, alg_format=\"fasta\")\n",
    "t = Tree(treefile, format=9)\n",
    "#t.ladderize(direction=1)\n",
    "leafSet = set()\n",
    "\n",
    "#print(\"Custom mode:\")\n",
    "for n in t.get_leaves():\n",
    "    leafSet.add(n.name)\n",
    "    #print(n.name)\n",
    "    tmp = n.name.split(\"_\") \n",
    "    spid = tmp.pop(-1)\n",
    "    tmp.pop(-1)\n",
    "    speciesname = code2name[spid]\n",
    "    genename = \"_\".join(tmp)\n",
    "    #print(speciesname, genename, code2lineage[spid])\n",
    "    \n",
    "    n.add_features(lineage=code2lineage[spid])\n",
    "    n.add_features(gene=genename)\n",
    "    n.add_features(species=speciesname)\n",
    "\n",
    "    #print(\"spid:\", spid, \"Species name:\", n.species, \"Species lineage:\", n.lineage, \"Color:\", lin2color[n.lineage])\n",
    "\n",
    "    # create a new label with a color attribute\n",
    "    linF = AttrFace(\"lineage\", fgcolor=lin2color[n.lineage], fsize=1)\n",
    "    linF.background.color = lin2color[n.lineage]\n",
    "    linF.margin_top = linF.margin_bottom = linF.margin_left = 10\n",
    "    \n",
    "    speciesF = AttrFace(\"species\", fsize=10, fgcolor=lin2color[n.lineage], fstyle=\"italic\")\n",
    "    speciesF.margin_right = speciesF.margin_left = 10\n",
    "\n",
    "    if spid == 'cri':\n",
    "        geneF = AttrFace(\"gene\", fsize=12, fgcolor=\"#228833\", fstyle=\"bold\")\n",
    "        geneF.margin_right = geneF.margin_left = 5\n",
    "    \n",
    "    elif spid == 'ath':\n",
    "        geneF = AttrFace(\"gene\", fsize=12, fgcolor=\"#AA3377\", fstyle=\"bold\")\n",
    "        geneF.margin_right = geneF.margin_left = 5\n",
    "    \n",
    "    else:\n",
    "        geneF = AttrFace(\"gene\", fsize=10, fgcolor=\"black\")\n",
    "        geneF.margin_right = geneF.margin_left = 5\n",
    "\n",
    "    # labels aligned to the same level\n",
    "    n.add_face(speciesF, 0, position='aligned')\n",
    "    n.add_face(geneF, 0, position='branch-right')\n",
    "    n.add_face(linF, 1, position='aligned')\n",
    "    \n",
    "    #my_motifs = [[0, alnLen, \"seq\", 2, 10, None, None, None]]\n",
    "    #seqF = SeqMotifFace(seq=seqDict[n.name], motifs=my_motifs, gap_format=\"blank\")\n",
    "    #seqF.margin_right = seqF.margin_left = 5\n",
    "    #n.add_face(seqF, 2, \"aligned\")\n",
    "    \n",
    "    seqF = SeqMotifFace(seqDict[n.name][947:1026], seq_format=\"seq\", height=22)\n",
    "    seqF.margin_right = seqF.margin_left = 5\n",
    "    #seqF.margin_top = seqF.margin_bottom = 0\n",
    "    n.add_face(seqF, 2, \"aligned\")\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "53e12b1c",
   "metadata": {},
   "outputs": [
    {
     "ename": "IndexError",
     "evalue": "string index out of range",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mIndexError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-69-c509fa02a40b>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     38\u001b[0m     \u001b[0;31m#fix branchlengths?\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     39\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 40\u001b[0;31m     \u001b[0;32mif\u001b[0m \u001b[0mn\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mname\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mleafSet\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0mn\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0;34m'n'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     41\u001b[0m         \u001b[0;31m#print(n.name)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     42\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mIndexError\u001b[0m: string index out of range"
     ]
    }
   ],
   "source": [
    "# add lineage information to all internal nodes\n",
    "style = NodeStyle()\n",
    "\n",
    "style[\"size\"] = 0\n",
    "style[\"hz_line_width\"] = 2\n",
    "style[\"vt_line_width\"] = 2\n",
    "t.set_style(style)\n",
    "\n",
    "for n in t.iter_descendants(\"postorder\"):\n",
    "    #print(n.name)\n",
    "                \n",
    "    style[\"size\"] = 0\n",
    "    style[\"hz_line_width\"] = 2\n",
    "    style[\"vt_line_width\"] = 2\n",
    "    n.set_style(style)\n",
    "    \n",
    "    lineage_set = set()\n",
    "    # get descendants, if all descendants are members of same lineage, color lineage color\n",
    "    #print(\"NODE CHILDREN:\")\n",
    "    for k in n.iter_descendants(\"postorder\"):\n",
    "        for l in k.get_leaves():\n",
    "            lineage_set.add(l.lineage)\n",
    "            #print(\"Gene:\", l.gene, \"Species:\", l.species, \"Lineage:\", l.lineage, \"Color:\", lin2color[l.lineage])\n",
    "    \n",
    "    #print(len(lineage_set), lineage_set)\n",
    "    if len(lineage_set) == 1:\n",
    "        node_lin = ''.join(lineage_set)\n",
    "        #print(len(lineage_set), lineage_set, node_lin, lin2color[node_lin])\n",
    "    \n",
    "        newstyle = NodeStyle()\n",
    "        newstyle[\"size\"] = 0\n",
    "        newstyle[\"hz_line_width\"] = 2\n",
    "        newstyle[\"vt_line_width\"] = 2\n",
    "        newstyle[\"vt_line_color\"] = lin2color[node_lin]\n",
    "        newstyle[\"hz_line_color\"] = lin2color[node_lin]\n",
    "        n.img_style = newstyle\n",
    "        \n",
    "    #fix branchlengths?\n",
    "    \n",
    "    if n.name not in leafSet and n.name[0] != 'n':\n",
    "        #print(n.name)\n",
    "        \n",
    "        if float(n.name) >= branch_min:\n",
    "            #branch_support = n.name.split('.')[0]\n",
    "            branch_support = n.name\n",
    "            #print(branch_support)\n",
    "            n.add_features(bootstrap=branch_support)\n",
    "            \n",
    "            if len(lineage_set) == 1:\n",
    "                node_lin = ''.join(lineage_set)\n",
    "                supF = AttrFace(\"bootstrap\", fgcolor=lin2color[node_lin], fsize=8)\n",
    "                supF.margin_right = supF.margin_left = 3\n",
    "                n.add_face(supF, 0, position='branch-bottom')\n",
    "                \n",
    "            else:\n",
    "                supF = AttrFace(\"bootstrap\", fgcolor=\"#000000\", fsize=8)\n",
    "                supF.margin_right = supF.margin_left = 3\n",
    "                n.add_face(supF, 0, position='branch-bottom')\n",
    "        \n",
    "\n",
    "for n in t.get_leaves():\n",
    "    \n",
    "    leafstyle = NodeStyle()\n",
    "    leafstyle[\"size\"] = 0\n",
    "    leafstyle[\"hz_line_width\"] = 2\n",
    "    leafstyle[\"vt_line_width\"] = 2\n",
    "    leafstyle[\"vt_line_color\"] = lin2color[n.lineage]\n",
    "    leafstyle[\"hz_line_color\"] = lin2color[n.lineage]\n",
    "    n.img_style = leafstyle\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "5e87ebff",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add legend\n",
    "\n",
    "ts = TreeStyle()\n",
    "ts.show_leaf_name = False\n",
    "#ts.show_branch_support = True\n",
    "ts.draw_guiding_lines = True\n",
    "\n",
    "ts.title.add_face(TextFace(\"Taxonomy:\", fsize=10), column=0)\n",
    "for i in range(1, len(lin2color)+1):\n",
    "    #print(linorder[i], lin2color[linorder[i]])\n",
    "    ts.title.add_face(TextFace(linorder[i], fsize=10, fgcolor=lin2color[linorder[i]]), column=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "c3d16346",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'nodes': [[45.5, 56.75, 49.5, 60.75, 1, None],\n",
       "  [93.5, 26.5, 97.5, 30.5, 2, None],\n",
       "  [93.5, 87.0, 97.5, 91.0, 7, None],\n",
       "  [141.5, 76.0, 145.5, 80.0, 8, None],\n",
       "  [141.5, 98.0, 145.5, 102.0, 9, None],\n",
       "  [45.5, 180.5, 49.5, 184.5, 10, None],\n",
       "  [93.5, 153.0, 97.5, 157.0, 11, None],\n",
       "  [141.5, 131.0, 145.5, 135.0, 12, None],\n",
       "  [189.5, 120.0, 193.5, 124.0, 13, None],\n",
       "  [189.5, 142.0, 193.5, 146.0, 14, None],\n",
       "  [141.5, 175.0, 145.5, 179.0, 15, None],\n",
       "  [189.5, 164.0, 193.5, 168.0, 16, None],\n",
       "  [189.5, 186.0, 193.5, 190.0, 17, None],\n",
       "  [93.5, 208.0, 97.5, 212.0, 18, None]],\n",
       " 'faces': [[147.0, 1.0, 263.0, 23.0, 3, 'AT4G39400.1'],\n",
       "  [554.5, 0.5, 575.5, 23.5, 3, None],\n",
       "  [396.0, 3.0, 538.0, 21.0, 3, 'Arabidopsis thaliana'],\n",
       "  [565.0, 11.0, 571.0, 13.0, 3, 'Angiosperm'],\n",
       "  [579.5, 0.5, 1370.5, 23.5, 3, ''],\n",
       "  [192.0, 23.0, 308.0, 45.0, 5, 'AT3G13380.1'],\n",
       "  [554.5, 22.5, 575.5, 45.5, 5, None],\n",
       "  [396.0, 25.0, 538.0, 43.0, 5, 'Arabidopsis thaliana'],\n",
       "  [565.0, 33.0, 571.0, 35.0, 5, 'Angiosperm'],\n",
       "  [579.5, 22.5, 1370.5, 45.5, 5, ''],\n",
       "  [192.0, 45.0, 308.0, 67.0, 6, 'AT1G55610.2'],\n",
       "  [554.5, 44.5, 575.5, 67.5, 6, None],\n",
       "  [396.0, 47.0, 538.0, 65.0, 6, 'Arabidopsis thaliana'],\n",
       "  [565.0, 55.0, 571.0, 57.0, 6, 'Angiosperm'],\n",
       "  [579.5, 44.5, 1370.5, 67.5, 6, ''],\n",
       "  [150.0, 67.0, 333.0, 89.0, 8, 'Ceric.29G061800.1.p'],\n",
       "  [554.5, 66.5, 575.5, 89.5, 8, None],\n",
       "  [396.0, 69.0, 545.0, 87.0, 8, 'Ceratopteris richardii'],\n",
       "  [565.0, 77.0, 575.0, 79.0, 8, 'Homosporous_Fern'],\n",
       "  [579.5, 66.5, 1370.5, 89.5, 8, ''],\n",
       "  [150.0, 89.0, 333.0, 111.0, 9, 'Ceric.21G054300.1.p'],\n",
       "  [554.5, 88.5, 575.5, 111.5, 9, None],\n",
       "  [396.0, 91.0, 545.0, 109.0, 9, 'Ceratopteris richardii'],\n",
       "  [565.0, 99.0, 575.0, 101.0, 9, 'Homosporous_Fern'],\n",
       "  [579.5, 88.5, 1370.5, 111.5, 9, ''],\n",
       "  [198.0, 111.0, 381.0, 133.0, 13, 'Ceric.04G066800.1.p'],\n",
       "  [554.5, 110.5, 575.5, 133.5, 13, None],\n",
       "  [396.0, 113.0, 545.0, 131.0, 13, 'Ceratopteris richardii'],\n",
       "  [565.0, 121.0, 575.0, 123.0, 13, 'Homosporous_Fern'],\n",
       "  [579.5, 110.5, 1370.5, 133.5, 13, ''],\n",
       "  [198.0, 133.0, 381.0, 155.0, 14, 'Ceric.38G044500.1.p'],\n",
       "  [554.5, 132.5, 575.5, 155.5, 14, None],\n",
       "  [396.0, 135.0, 545.0, 153.0, 14, 'Ceratopteris richardii'],\n",
       "  [565.0, 143.0, 575.0, 145.0, 14, 'Homosporous_Fern'],\n",
       "  [579.5, 132.5, 1370.5, 155.5, 14, ''],\n",
       "  [198.0, 155.0, 381.0, 177.0, 16, 'Ceric.15G059000.1.p'],\n",
       "  [554.5, 154.5, 575.5, 177.5, 16, None],\n",
       "  [396.0, 157.0, 545.0, 175.0, 16, 'Ceratopteris richardii'],\n",
       "  [565.0, 165.0, 575.0, 167.0, 16, 'Homosporous_Fern'],\n",
       "  [579.5, 154.5, 1370.5, 177.5, 16, ''],\n",
       "  [198.0, 177.0, 381.0, 199.0, 17, 'Ceric.19G058600.1.p'],\n",
       "  [554.5, 176.5, 575.5, 199.5, 17, None],\n",
       "  [396.0, 179.0, 545.0, 197.0, 17, 'Ceratopteris richardii'],\n",
       "  [565.0, 187.0, 575.0, 189.0, 17, 'Homosporous_Fern'],\n",
       "  [579.5, 176.5, 1370.5, 199.5, 17, ''],\n",
       "  [102.0, 199.0, 218.0, 221.0, 18, 'AT2G01950.1'],\n",
       "  [554.5, 198.5, 575.5, 221.5, 18, None],\n",
       "  [396.0, 201.0, 538.0, 219.0, 18, 'Arabidopsis thaliana'],\n",
       "  [565.0, 209.0, 571.0, 211.0, 18, 'Angiosperm'],\n",
       "  [579.5, 198.5, 1370.5, 221.5, 18, '']],\n",
       " 'node_areas': {0: [1.0, 1.0, 386.0, 221.0],\n",
       "  1: [1.0, 1.0, 338.0, 111.0],\n",
       "  2: [49.0, 1.0, 313.0, 67.0],\n",
       "  3: [97.0, 1.0, 268.0, 23.0],\n",
       "  4: [97.0, 23.0, 313.0, 67.0],\n",
       "  5: [142.0, 23.0, 313.0, 45.0],\n",
       "  6: [142.0, 45.0, 313.0, 67.0],\n",
       "  7: [49.0, 67.0, 338.0, 111.0],\n",
       "  8: [97.0, 67.0, 338.0, 89.0],\n",
       "  9: [97.0, 89.0, 338.0, 111.0],\n",
       "  10: [1.0, 111.0, 386.0, 221.0],\n",
       "  11: [49.0, 111.0, 386.0, 199.0],\n",
       "  12: [97.0, 111.0, 386.0, 155.0],\n",
       "  13: [145.0, 111.0, 386.0, 133.0],\n",
       "  14: [145.0, 133.0, 386.0, 155.0],\n",
       "  15: [97.0, 155.0, 386.0, 199.0],\n",
       "  16: [145.0, 155.0, 386.0, 177.0],\n",
       "  17: [145.0, 177.0, 386.0, 199.0],\n",
       "  18: [49.0, 199.0, 223.0, 221.0]}}"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# render image on notebook or save to file\n",
    "t.render(\"/home/jwisecav/mytree.pdf\", tree_style=ts)\n",
    "#t.render(\"%%inline\", tree_style=ts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b39de418",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d19177ea",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (My env.genomics Kernel)",
   "language": "python",
   "name": "env.genomics"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
