{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "98a408b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import Bio\n",
    "from Bio import Seq, SeqIO\n",
    "import ete3\n",
    "from ete3 import Tree, faces, TreeStyle, PhyloTree, NodeStyle, TextFace, AttrFace, SeqMotifFace"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b707db9b",
   "metadata": {},
   "outputs": [],
   "source": [
    "mltreefile = '../../bsa_by_rnaseq/05_phylogeny/CrBRL5.treefile'\n",
    "fintreefile = '../../bsa_by_rnaseq/05_phylogeny/generax/results/CrBRL5/geneTree.newick'\n",
    "outtreefile = '../../bsa_by_rnaseq/05_phylogeny/generax/results/CrBRL5/geneTree.fixboot.tre'\n",
    "alnfile = '../../bsa_by_rnaseq/05_phylogeny/CrBRL5.aln'\n",
    "speciesfile = 'species_ids.txt'\n",
    "colorfile = 'lineage_colors.dmp'\n",
    "branch_min = 95\n",
    "outfile = '../../bsa_by_rnaseq/05_phylogeny/generax/results/CrBRL5/geneTree.fixboot.pdf'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "283bb6fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Populate lookup dictionaries\n",
    "seqDict = {}\n",
    "alnLen = 0\n",
    "for record in SeqIO.parse(alnfile, \"fasta\"):\n",
    "    name = record.id\n",
    "    sequence = str(record.seq)\n",
    "    alnLen = len(sequence)\n",
    "    seqDict[name] = sequence\n",
    "\n",
    "code2name = {}\n",
    "code2lineage = {}\n",
    "fi = open(speciesfile)\n",
    "\n",
    "for line in fi:\n",
    "    if line[0] == '#':\n",
    "        continue\n",
    "    name, spid, lineage = line.rstrip().split('\\t')\n",
    "    #print(name, spid, lineage)\n",
    "    code2name[spid] = name\n",
    "    \n",
    "    code2lineage[spid] = lineage\n",
    "\n",
    "fi.close()\n",
    "\n",
    "lin2color = {}\n",
    "linorder = {}\n",
    "fi = open(colorfile)\n",
    "\n",
    "for line in fi:\n",
    "    if line[0] == '#':\n",
    "        continue\n",
    "\n",
    "    order, color_desc, color, cat, lineage, taxids = line.rstrip().split('\\t')  \n",
    "    order = int(order)\n",
    "    lin2color[lineage] = color\n",
    "    linorder[order] = lineage\n",
    "\n",
    "fi.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "fd1c8df3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# store support values from ML tree in lookup table\n",
    "mlt = Tree(mltreefile, format=1)\n",
    "\n",
    "nodeDict = {}\n",
    "leafSet = set()\n",
    "\n",
    "for n in mlt.get_leaves():\n",
    "    leafSet.add(n.name)\n",
    "\n",
    "for n in mlt.iter_descendants(\"postorder\"):\n",
    "    nodeSet = set()\n",
    "    # get descendants, if all descendants are members of same lineage, color lineage color\n",
    "    #print(\"NODE CHILDREN:\")\n",
    "    for k in n.iter_descendants(\"postorder\"):\n",
    "        for l in k.get_leaves():\n",
    "            #print(l.name)\n",
    "            nodeSet.add(l.name)\n",
    "            #print(\"Gene:\", l.gene, \"Species:\", l.species, \"Lineage:\", l.lineage, \"Color:\", lin2color[l.lineage])\n",
    "    if n.name not in leafSet and n.name != '':\n",
    "        #print(n.name)\n",
    "        branch1, branch2 = n.name.split('/')\n",
    "        \n",
    "        nodeDict[frozenset(nodeSet)] = [branch1, branch2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "1681241e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add species and lineage information to all leaves\n",
    "t = Tree(fintreefile, format=1)\n",
    "t.ladderize(direction=1)\n",
    "leafSet = set()\n",
    "\n",
    "for n in t.get_leaves():\n",
    "    leafSet.add(n.name)\n",
    "    #print(n.name)\n",
    "    tmp = n.name.split(\"_\") \n",
    "    spid = tmp.pop(-1)\n",
    "    tmp.pop(-1)\n",
    "    speciesname = code2name[spid]\n",
    "    genename = \"_\".join(tmp)\n",
    "    #print(speciesname, genename, code2lineage[spid])\n",
    "    \n",
    "    n.add_features(lineage=code2lineage[spid])\n",
    "    n.add_features(gene=genename)\n",
    "    n.add_features(species=speciesname)\n",
    "\n",
    "    #print(\"spid:\", spid, \"Species name:\", n.species, \"Species lineage:\", n.lineage, \"Color:\", lin2color[n.lineage])\n",
    "\n",
    "    # create a new label with a color attribute\n",
    "    linF = AttrFace(\"lineage\", fgcolor=lin2color[n.lineage], fsize=1)\n",
    "    linF.background.color = lin2color[n.lineage]\n",
    "    linF.margin_top = linF.margin_bottom = linF.margin_left = 10\n",
    "    \n",
    "    speciesF = AttrFace(\"species\", fsize=10, fgcolor=lin2color[n.lineage], fstyle=\"italic\")\n",
    "    speciesF.margin_right = speciesF.margin_left = 10\n",
    "\n",
    "    if spid == 'cri':\n",
    "        geneF = AttrFace(\"gene\", fsize=12, fgcolor=\"#e31a1c\", fstyle=\"bold\")\n",
    "        geneF.margin_right = geneF.margin_left = 5\n",
    "    \n",
    "    elif spid == 'ath':\n",
    "        geneF = AttrFace(\"gene\", fsize=12, fgcolor=\"#e31a1c\", fstyle=\"bold\")\n",
    "        geneF.margin_right = geneF.margin_left = 5\n",
    "    \n",
    "    else:\n",
    "        geneF = AttrFace(\"gene\", fsize=10, fgcolor=\"black\")\n",
    "        geneF.margin_right = geneF.margin_left = 5\n",
    "\n",
    "    # labels aligned to the same level\n",
    "    n.add_face(speciesF, 0, position='aligned')\n",
    "    n.add_face(geneF, 0, position='branch-right')\n",
    "    n.add_face(linF, 1, position='aligned')\n",
    "    \n",
    "    my_motifs = [[0, alnLen, \"compactseq\", 2, 10, None, None, None]]\n",
    "    seqF = SeqMotifFace(seq=seqDict[n.name], motifs=my_motifs, gap_format=\"blank\")\n",
    "    seqF.margin_right = seqF.margin_left = 5\n",
    "    n.add_face(seqF, 2, \"aligned\")\n",
    "    \n",
    "\n",
    "# add lineage information to all internal nodes\n",
    "style = NodeStyle()\n",
    "\n",
    "style[\"size\"] = 0\n",
    "style[\"hz_line_width\"] = 2\n",
    "style[\"vt_line_width\"] = 2\n",
    "t.set_style(style)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "93982f85",
   "metadata": {},
   "outputs": [],
   "source": [
    "for n in t.iter_descendants(\"postorder\"):\n",
    "    #print(n.name)\n",
    "                \n",
    "    style[\"size\"] = 0\n",
    "    style[\"hz_line_width\"] = 2\n",
    "    style[\"vt_line_width\"] = 2\n",
    "    n.set_style(style)\n",
    "    \n",
    "    lineage_set = set()\n",
    "    nodeSet = set()\n",
    "    # get descendants, if all descendants are members of same lineage, color lineage color\n",
    "    #print(\"NODE CHILDREN:\")\n",
    "    for k in n.iter_descendants(\"postorder\"):\n",
    "        for l in k.get_leaves():\n",
    "            lineage_set.add(l.lineage)\n",
    "            nodeSet.add(l.name)\n",
    "            #print(\"Gene:\", l.gene, \"Species:\", l.species, \"Lineage:\", l.lineage, \"Color:\", lin2color[l.lineage])\n",
    "    \n",
    "    #print(len(lineage_set), lineage_set)\n",
    "    if len(lineage_set) == 1:\n",
    "        node_lin = ''.join(lineage_set)\n",
    "        #print(len(lineage_set), lineage_set, node_lin, lin2color[node_lin])\n",
    "    \n",
    "        newstyle = NodeStyle()\n",
    "        newstyle[\"size\"] = 0\n",
    "        newstyle[\"hz_line_width\"] = 2\n",
    "        newstyle[\"vt_line_width\"] = 2\n",
    "        newstyle[\"vt_line_color\"] = lin2color[node_lin]\n",
    "        newstyle[\"hz_line_color\"] = lin2color[node_lin]\n",
    "        n.img_style = newstyle\n",
    "        \n",
    "    #fix branchlengths    \n",
    "    if n.name not in leafSet and n.name != '':\n",
    "        #print(n.name)\n",
    "\n",
    "        if frozenset(nodeSet) in nodeDict:\n",
    "            branch1 = nodeDict[frozenset(nodeSet)][0]\n",
    "            branch2 = nodeDict[frozenset(nodeSet)][1]\n",
    "            #print(branch2)\n",
    "\n",
    "            n.add_features(bootstrap1=branch1)\n",
    "            n.add_features(bootstrap2=branch2)\n",
    "\n",
    "            if len(lineage_set) == 1:\n",
    "                node_lin = ''.join(lineage_set)\n",
    "                supF = AttrFace(\"bootstrap1\", fgcolor=lin2color[node_lin], fsize=8)\n",
    "                supF.margin_right = supF.margin_left = 3\n",
    "                n.add_face(supF, 0, position='branch-top')\n",
    "\n",
    "                supF = AttrFace(\"bootstrap2\", fgcolor=lin2color[node_lin], fsize=8)\n",
    "                supF.margin_right = supF.margin_left = 3\n",
    "                n.add_face(supF, 0, position='branch-bottom')\n",
    "\n",
    "            else:\n",
    "                supF = AttrFace(\"bootstrap1\", fgcolor=\"#000000\", fsize=8)\n",
    "                supF.margin_right = supF.margin_left = 3\n",
    "                n.add_face(supF, 0, position='branch-top')\n",
    "\n",
    "                supF = AttrFace(\"bootstrap2\", fgcolor=\"#000000\", fsize=8)\n",
    "                supF.margin_right = supF.margin_left = 3\n",
    "                n.add_face(supF, 0, position='branch-bottom')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d28e1a81",
   "metadata": {},
   "outputs": [],
   "source": [
    "for n in t.get_leaves():\n",
    "    \n",
    "    leafstyle = NodeStyle()\n",
    "    leafstyle[\"size\"] = 0\n",
    "    leafstyle[\"hz_line_width\"] = 2\n",
    "    leafstyle[\"vt_line_width\"] = 2\n",
    "    leafstyle[\"vt_line_color\"] = lin2color[n.lineage]\n",
    "    leafstyle[\"hz_line_color\"] = lin2color[n.lineage]\n",
    "    n.img_style = leafstyle\n",
    "    \n",
    "    \n",
    "ts = TreeStyle()\n",
    "ts.show_leaf_name = False\n",
    "ts.draw_guiding_lines = True\n",
    "\n",
    "# add legend\n",
    "ts.title.add_face(TextFace(\"Taxonomy:\", fsize=10), column=0)\n",
    "for i in range(1, len(lin2color)+1):\n",
    "    #print(linorder[i], lin2color[linorder[i]])\n",
    "    ts.title.add_face(TextFace(linorder[i], fsize=10, fgcolor=lin2color[linorder[i]]), column=0)\n",
    "\n",
    "\n",
    "# render image on notebook or save to file\n",
    "#t.render(outfile, tree_style=ts)\n",
    "t.render(\"%%inline\", tree_style=ts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40c9e59e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (My env.genomics Kernel)",
   "language": "python",
   "name": "env.genomics"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
